Handle unparsed non-optional complex property types

gwynne opened the pull request:
As initially demonstrated in #553, the existing `ArgumentDecoder` logic fails when presented with a property which:

- Is unparsed
- Is not optional
- Has a default value, and
- Requests a keyed container during decoding

This occurs because while the actual property in question is properly assessed as having `.defaultValue` as an `InputOrigin`, the request for a keyed decoding container bypasses the logic which would otherwise return said default value, and since the `ArgumentDefinition` logic is (correctly) not recursive, the type fails to decode any keys it contains due to a "missing value".

The most straightforward solution I could find, one which covers all cases without negatively affecting handling of parsed properties (to the best of my ability to determine), was to apply the "default value" handling from the `decodeIfPresent<T>(_:forKey:)` method of the keyed decoding container (which ensured that _optional_ properties of this sort already worked correctly) to the more fundamental "bottleneck" `decode<T>(_:forKey:)` method as well, modified to throw a `wrongType` error instead of returning nil on type mismatch.

I additionally added a functioning `allKeys` property accessor for additional robustness against custom `Decodable.init(from:)` implementations. (The choice not to do the same for `nestedContainer(keyedBy:)`, `nestedUnkeyedContainer()`, `superDecoder()`, and `superDecoder(forKey:)` was simply pragmatic - it's not even close to worth the effort.)

Fixes #553 

### Checklist
- [x] I've added at least one test that validates that my change is working, if appropriate
- [x] I've followed the code style of the rest of the project
- [x] I've read the [Contribution Guidelines](https://github.com/apple/swift-argument-parser/blob/main/CONTRIBUTING.md)
- [x] I've updated the documentation if necessary

natecook1000 reviewed (commented):
@gwynne Thanks so much for this fix! I believe your change to `decode(_:forKey:)` is sufficient to solve the problem. Unless we can create a test that exercises the new `allKeys`, let's revert that back to a fatal error.

Can you also switch to 2-space indenting? Thanks!

gwynne commented:
@natecook1000 Well, I tried to come up with a test that would exercise the `allKeys` implementation, but it turns out nothing I could think of would get that far with the change to `decode(_:forKey:)` in place; I've removed it and put back the `fatalError()` - and also fixed my indentation, don't worry! 😆

natecook1000 commented:
@swift-ci Please test

hassila commented:
(thanks @gwynne for fixing this, much obliged!)
